File: src/chipschallenge/blockreactions/BlueWallRealTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Block.Type;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Move;
import chipschallenge.Move.Moves;

/**
 *
 * @author patrik
 */
public class BlueWallRealTo extends BlockReaction {

    private BlueWallRealTo() {}
    private static BlueWallRealTo mInstance = null;
    public static synchronized BlueWallRealTo getInstance() {
        if(mInstance == null)
            mInstance = new BlueWallRealTo();
        return mInstance;
    }

    public void react(Block moving, Block standing) throws BlockContainerFullException {
        Moves facing = moving.getFacing();
        moving.move(Move.reverse(facing));
        moving.setFacing(facing);
        standing.replace(createBlock(Type.WALL));
    }

    public boolean canMove(Block moving, Block standing) {
        return moving.isChip();
    }

}

File: ChipsChallenge.Shared/Blockreactions/BlueWallRealTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    public class BlueWallRealTo : NoSlipReaction
    {
        private static BlueWallRealTo instance;

        private BlueWallRealTo()
        {
        }

        public static BlueWallRealTo Instance
        {
            get
            {
                lock (typeof(BlueWallRealTo))
                {
                    return instance ?? (instance = new BlueWallRealTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            Move.Moves facing = moving.Facing;
            moving.Move(Move.Reverse(facing));
            moving.Facing = facing;
            standing.Replace(CreateBlock(Block.Type.WALL));
            Sound().Play(Shared.Sound.ChipHum);
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip;
        }
    }
}